### PR TITLE
Add CI Workflow for Building and Deploying Operator

### PR DESCRIPTION
**Pull request by Parthiba-Hazra**
- Add GitHub Actions workflow for building and deploying the checkpoint-restore-operator to the kind kubernetes cluster
 - Add verification workflows for linting and go module checks.

### CODE REVIEW

**Comment by Parthiba-Hazra** [comment]
May be assuring only the ready state of **checkpoint-restore-operator-controller-manager** deployment is not enough. Let me know any other things we need to assure for the operator deployment.

File: .github/workflows/quickstart.yaml [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+                    kubectl -n checkpoint-restore-operator-system get pods
+                    kubectl -n checkpoint-restore-operator-system describe pods
+                    exit 1
+                fi
+                echo "Waiting for ${deployment_name} deployment to report a ready status"
+                sleep 5
+            done
+            echo "The ${deployment_name} deployment is ready"
+          }
+          wait_for_deployment checkpoint-restore-operator-controller-manager
```
Comment (snprajwal) [inline .github/workflows/quickstart.yaml:74]: It would be better to have a script under `test` with these contents and invoke the script from the CI. Going forward, if we decide to make this any more complicated, the YAML will become unreadable.

File: .github/workflows/quickstart.yaml [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        uses: actions/checkout@v4
+
+      - name: Set up Go
+        uses: actions/setup-go@v4
+        with:
+          go-version: 1.21
+
+      - name: Install dependencies
+        run: |
+          # Install kind
```
Comment (snprajwal) [inline .github/workflows/quickstart.yaml:19]: Would it make more sense to use https://github.com/helm/kind-action instead of manually doing this ourselves?

File: .github/workflows/quickstart.yaml (hunk 1 above)
Comment (Parthiba-Hazra) [reply .github/workflows/quickstart.yaml:74]: Got it, but what about this - 

> May be assuring only the ready state of checkpoint-restore-operator-controller-manager deployment is not 
> enough. Let me know any other things we need to assure for the operator deployment.

any thoughts?

File: .github/workflows/quickstart.yaml [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        run: |
+          kind load docker-image checkpoint-restore-operator:ci
+
+      - name: Deploy to Kubernetes
+        run: |
+          make install
+          make deploy IMG=checkpoint-restore-operator:ci
+
+      - name: Add execute permissions to wait_for_deployment.sh
+        run: chmod +x ./test/wait_for_deployment.sh
```
Comment (rst0git) [inline .github/workflows/quickstart.yaml:52]: Would it be possible to commit this file with execute permissions, e.g., `chmod +x ... && git commit -a --amend`?

File: .github/workflows/verify.yaml [hunk 4]
```diff
@@ -0,0 +1,70 @@
+name: verify
```
Comment (rst0git) [inline .github/workflows/verify.yaml:1]: This should be in a separate commit.

File: .github/dependabot.yaml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+version: 2
```
Comment (rst0git) [inline .github/dependabot.yaml:1]: This should be a separate commit adding dependabot configuration.

File: Makefile [hunk 6]
```diff
@@ -121,6 +121,12 @@ test: manifests generate fmt vet envtest ## Run tests.
 build: manifests generate fmt vet ## Build manager binary.
 	go build -o bin/manager cmd/main.go
 
+.PHONY: vendor
+vendor:
+	go mod tidy
+	go mod vendor
+	go mod verify
```
Comment (rst0git) [inline Makefile:128]: This should be in a separate commit adding only the `vendor` target to the Makefile.

File: .github/workflows/quickstart.yaml (hunk 1 above)
Comment (rst0git) [reply .github/workflows/quickstart.yaml]: > any thoughts?

This is a good starting point. We can add more comprehensive tests later in another pull request.

**Comment by adrianreber** [comment]
Is this deployment using containerd or CRI-O. For now we would need to use CRI-O to be able to use the checkpointing feature.

**Comment by rst0git** [comment]
@adrianreber I'm not sure if we should use the checkpointing feature. It would be easier to use a script that generates a set of checkpoint files (tar archives) similar to the approach we used with checkpointctl.

**Comment by Parthiba-Hazra** [comment]
> Is this deployment using containerd or CRI-O. For now we would need to use CRI-O to be able to use the checkpointing feature.

I think its containered and i did not installed CRI-O as I am not sure if checkpointing in CI would be a good idea. So, mainly here we are only asserting if operator deployed successfully or not. 

So according to you we should also assert that the manager applying policies on checkpoints correctly.

**Comment by adrianreber** [comment]
> > Is this deployment using containerd or CRI-O. For now we would need to use CRI-O to be able to use the checkpointing feature.
> 
> I think its containered and i did not installed CRI-O as I am not sure if checkpointing in CI would be a good idea. So, mainly here we are only asserting if operator deployed successfully or not.

Makes sense.

> So according to you we should also assert that the manager applying policies on checkpoints correctly.

That could be a second step. As @rst0git says we could create a couple of files manually.

File: test/wait_for_deployment.sh [hunk 7]
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
```
Comment (adrianreber) [inline test/wait_for_deployment.sh:1]: For the shell script please also do a ShellCheck and shfmt CI target.

File: config/crd/bases/criu.org_checkpointrestoreoperators.yaml [hunk 8]
```diff
@@ -51,4 +51,4 @@ spec:
     served: true
     storage: true
     subresources:
-      status: {}
+      status: {}
```
Comment (kolyshkin) [inline config/crd/bases/criu.org_checkpointrestoreoperators.yaml:54]: useless change removing newline at EOF

File: .github/dependabot.yaml [hunk 9]
```diff
@@ -0,0 +1,7 @@
...
+updates:
+  # Check for updates to Go modules
+  - package-ecosystem: "gomod"
+    directory: "/"
+    schedule:
+      interval: "weekly"
```
Comment (kolyshkin) [inline .github/dependabot.yaml:7]: nit: missing newline at EOF

File: .github/workflows/verify.yaml [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+on: [push, pull_request]
+
+jobs:
+  lint:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+
+      - name: Set up Go
+        uses: actions/setup-go@v3
```
Comment (kolyshkin) [inline .github/workflows/verify.yaml:12]: latest is v5, please use it.

File: .github/workflows/verify.yaml [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+          go-version: '1.21'
+
+      - name: Install ShellCheck and shfmt
+        run: |
+          sudo apt update
+          sudo apt install -y shellcheck
+          go install mvdan.cc/sh/v3/cmd/shfmt@latest
+
+      - name: golangci-lint
+        uses: golangci/golangci-lint-action@v3
```
Comment (kolyshkin) [inline .github/workflows/verify.yaml:23]: latest is v6, please use it

File: .github/workflows/verify.yaml [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+        with:
+          version: latest
+          args: --timeout=5m --out-format=colored-line-number
```
Comment (kolyshkin) [inline .github/workflows/verify.yaml:26]: I _think_ that if you use a non-default format, gha won't be able to annotate linter errors.

File: .github/workflows/verify.yaml [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+      - name: codespell
+        uses: codespell-project/actions-codespell@v2
+        with:
+          skip: vendor,*.svg
+
+      - name: ShellCheck for wait_for_deployment.sh
+        run: shellcheck ./test/wait_for_deployment.sh
+
+      - name: shfmt for wait_for_deployment.sh
+        run: shfmt -d ./test/wait_for_deployment.sh
```
Comment (kolyshkin) [inline .github/workflows/verify.yaml:38]: why are there these two separate steps for one specific shell file?

File: .github/workflows/verify.yaml [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+
+  gomod:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - name: Set up Go
+        uses: actions/setup-go@v3
```
Comment (kolyshkin) [inline .github/workflows/verify.yaml:45]: see above

File: .github/workflows/verify.yaml [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+        run: |
+          make vendor
+          git diff --exit-code
+
+      - name: Upload go.sum diff if exists
+        if: failure()
+        uses: actions/upload-artifact@v3
+        with:
+          name: go-sum-diff
+          path: go.sum.diff
```
Comment (kolyshkin) [inline .github/workflows/verify.yaml:59]: no need to have this; CI job failure logs are enough to find out what's wrong.

Comment (Parthiba-Hazra) [reply .github/workflows/verify.yaml:59]: @kolyshkin thank you for your review. To be honest I haven't worked with GitHub workflows that much previously so I think did some mistakes here unintentionally.

File: .github/workflows/quickstart.yaml (hunk 2 above)
Comment (rst0git) [reply .github/workflows/quickstart.yaml:19]: I'm not sure if it would be possible to load a container image with the GitHub Action
```
kind load docker-image checkpoint-restore-operator:ci
```

File: .github/workflows/verify.yaml (hunk 11 above)
Comment (rst0git) [reply .github/workflows/verify.yaml]: @Parthiba-Hazra GitHub automatically creates a unique token in the workflow:
https://docs.github.com/en/actions/security-guides/automatic-token-authentication

File: .github/workflows/verify.yaml [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+
+      - name: codespell
+        uses: codespell-project/actions-codespell@v2
+        with:
+          skip: vendor,*.svg
+      
+      - name: ShellCheck and shfmt for wait_for_deployment.sh
+        run: |
+          shellcheck ./test/wait_for_deployment.sh
+          shfmt -d ./test/wait_for_deployment.sh
```
Comment (snprajwal) [inline .github/workflows/verify.yaml:39]:
```suggestion
      - name: ShellCheck and shfmt
        run:
          shellcheck ./test/*.sh
          shfmt -d ./test/*.sh
```
We should be running it on a general pattern since we may add more files going forward.

File: .github/workflows/quickstart.yaml [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+          chmod +x kubectl
+          sudo mv kubectl /bin/
+          
+      - name: Build Docker image
+        run: |
+          sudo -E docker build -t checkpoint-restore-operator:ci .
+
+      - name: Delete existing Kubernetes cluster (if any)
+        run: |
+          kind delete cluster || true
```
Comment (rst0git) [inline .github/workflows/quickstart.yaml:35]: This step looks unnecessary. We don't expect to have an existing Kubernetes cluster in CI.

File: .github/workflows/quickstart.yaml [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+          sudo mv kind /bin/
+          
+          # Install kubectl
+          curl -LO "https://storage.googleapis.com/kubernetes-release/release/$(curl -s https://storage.googleapis.com/kubernetes-release/release/stable.txt)/bin/linux/amd64/kubectl"
+          chmod +x kubectl
+          sudo mv kubectl /bin/
+          
+      - name: Build Docker image
+        run: |
+          sudo -E docker build -t checkpoint-restore-operator:ci .
```
Comment (rst0git) [inline .github/workflows/quickstart.yaml:31]: Could we combine build and load into one step?
```yaml
      - name: Load Docker image into kind cluster
        run: |
          sudo -E docker build -t checkpoint-restore-operator:ci .
          kind load docker-image checkpoint-restore-operator:ci
```

File: .github/workflows/quickstart.yaml [hunk 19]
```diff
@@ -0,0 +1,55 @@
+name: Build and Deploy
```
Comment (rst0git) [inline .github/workflows/quickstart.yaml:1]: Could you rename this file to something like `tests.yaml`?

File: .github/workflows/quickstart.yaml (hunk 18 above)
Comment (rst0git) [inline .github/workflows/quickstart.yaml:31]:
```Suggestion
          make docker-build IMG=checkpoint-restore-operator:ci
```

File: test/wait_for_deployment.sh [hunk 20]
```diff
@@ -0,0 +1,25 @@
...
+
+wait_for_deployment() {
+	local deployment_name="$1"
+	timeout=60
```
Comment (rst0git) [inline test/wait_for_deployment.sh:5]: Could you add a comment to clarify the timeout here is 5 minutes?
```bash
timeout=60  # 5 minutes (60 * 5 sec)
```

File: .github/workflows/verify.yaml [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+              - 'README.md'
+              - '.github/workflows/verify.yml'
+
+      - name: Lint markdown
+        if: steps.changes.outputs.md == 'true'
+        uses: DavidAnson/markdownlint-cli2-action@v10
+        with:
+          globs: |
+            README.md
+            .github/workflows/verify.yml
```
Comment (rst0git) [inline .github/workflows/verify.yaml:76]: It looks like [`glob`](https://github.com/DavidAnson/markdownlint-cli2-action?tab=readme-ov-file#globs-optional) is used to specify Markdown files to lint. Why is `verify.yml` added here?

Comment (Parthiba-Hazra) [reply .github/workflows/verify.yaml:76]: It's not really necessary I think I just add it thinking it may contain Markdown content that needs to be linted.

File: .github/workflows/verify.yaml [hunk 22]
```diff
@@ -0,0 +1,75 @@
...
+          go install mvdan.cc/sh/v3/cmd/shfmt@latest
+
+      - name: golangci-lint
+        uses: golangci/golangci-lint-action@v6
+        with:
+          version: latest
+          only-new-issues: true
+          args: --timeout=5m
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
Comment (snprajwal) [inline .github/workflows/verify.yaml:29]: I would say these lines are not necessary. The token itself is created by the runner for each job, and a repo this small wouldn't risk timing out (and if it does, then that would more likely be a bug in golangci-lint)

File: .github/workflows/verify.yaml [hunk 23]
```diff
@@ -0,0 +1,75 @@
...
+      - name: Set up Go
+        uses: actions/setup-go@v5
+        with:
+          go-version: '1.21'
+
+      - name: Install ShellCheck and shfmt
+        run: |
+          sudo apt update
+          sudo apt install -y shellcheck
+          go install mvdan.cc/sh/v3/cmd/shfmt@latest
```
Comment (snprajwal) [inline .github/workflows/verify.yaml:20]: Is there a specific reason why we're not installing this through apt along with shellcheck?

File: .github/workflows/verify.yaml (hunk 21 above)
Comment (snprajwal) [reply .github/workflows/verify.yaml]: @rst0git I think it was added to allow the job to be triggered when the workflow file itself is changed, even if no other markdown files are changed in the commit.

Comment (snprajwal) [reply .github/workflows/verify.yaml]: Never mind, I just realised I was looking at the job above this (the paths filtering one). My bad

File: .github/workflows/verify.yaml (hunk 22 above)
Comment (Parthiba-Hazra) [reply .github/workflows/verify.yaml:29]: I think the job was failing without the env(only for the v6 version)

File: .github/workflows/verify.yaml (hunk 23 above)
Comment (Parthiba-Hazra) [reply .github/workflows/verify.yaml:20]: I tried it but apt can't find the package.

Comment (snprajwal) [reply .github/workflows/verify.yaml:20]: That's strange, it definitely exists on the package repository for 22.04: https://packages.ubuntu.com/jammy/shfmt

Comment (Parthiba-Hazra) [reply .github/workflows/verify.yaml:20]: Let me check again .. maybe I was testing it locally and it was failing there.

Comment (Parthiba-Hazra) [reply .github/workflows/verify.yaml]: yes you are right its works fine

**Comment by rst0git** [comment]
@adrianreber PTAL

**Comment by adrianreber** [comment]
Looks good. Let's use this as a start.

**Review by rst0git — APPROVED** [review]
